src/util/sss_ini.c

```diff
@@ -621,6 +621,63 @@ static errno_t check_domain_inherit_from(char *cfg_section,
     return ret;
 }
 
+static errno_t check_domain_id_provider(char *cfg_section,
+                                        struct ini_cfgobj *config_obj,
+                                        struct ini_errobj *errobj)
+{
+    struct value_obj *vo = NULL;
+    const char *valid_values[] = { "ad", "files", "ipa", "ldap", "proxy", NULL };
+    const char **valid_value;
+    const char *value;
+    int ret;
+
+    ret = ini_get_config_valueobj(cfg_section,
+                                  "id_provider",
+                                  config_obj,
+                                  INI_GET_NEXT_VALUE,
+                                  &vo);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    if (vo == NULL) {
+        ret = ini_errobj_add_msg(errobj,
+                                 "Attribute 'id_provider' is "
+                                 "missing in section '%s'.",
+                                 cfg_section);
+    } else {
+        value = sss_ini_get_const_string_config_value(vo, &ret);
+        if (ret != EOK) {
+            goto done;
+        }
+
+        valid_value = valid_values;
+        while (*valid_value != NULL) {
+            if (strcmp(value, *valid_value) == 0) {
+                break;
+            }
+            valid_value++;
+        }
+        if (*valid_value == NULL) {
+            ret = ini_errobj_add_msg(errobj,
+                                     "Attribute 'id_provider' in section '%s' "
+                                     "has an invalid value: %s",
+                                     cfg_section, value);
+            if (ret != EOK) {
+                goto done;
+            }
+        }
+    }
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+#define SECTION_IS_DOMAIN(s)    (strncmp("domain/", s, strlen("domain/")) == 0)
+#define SECTION_IS_SUBDOMAIN(s) (strchr(s + strlen("domain/"), '/') != NULL)
```

Comment:
First of all, please let's move definitions to the top of the file. Or even better, to a generic file that can be reused across the project.

Second, if the purpose of the macros is to differentiate between domain and subdomain, and taking into account that it can be one or the other (XOR). Then, SECTION_IS_DOMAIN would yield true for both a domain and a subdomain input. So, I think that you should change it to check the contrary of SECTION_IS_SUBDOMAIN. Example:  `strchr(s + strlen("domain/"), '/') == NULL`.

Edit: now that I've read all the code in this PR my previous comment seems out of context. Still, I'd like to stress that my understanding is that a section can be either a domain or a subdomain (or some other like application domains), and that's why I think the SECTION_IS_DOMAIN isn't correct. I'm to some discussion though.

Reply:
These two macros work together. `SECTION_IS_DOMAIN()` tells whether a section is a domain (start with "domain/") while `SECTION_IS_SUBDOMAIN()` is applicable to sections that we already know are domains.  I used `s + strlen("domain/")` not caring to check whether the string actually starts with "domain/." because of this precondition. Their use is quite specific to this module (actually specific to the function, that's why they are right above it), so I don't think it is a good idea to move them to a generic header file. To do so they will have to be re-written to behave correctly on every possible case. With this in mind, it should now be clear that their behavior is not XOR. May I should rename `SECTION_IS_SUBDOMAIN()` to `SECTION_DOMAIN_IS_SUBDOMAIN()`? Maybe this is not how the rest of the team understand it? A subdomain in not a domain?

If you still think they need to be in a generic header file for everyone to use, I will make them stronger.

On the other hand, I can simply remove them as they exist only to make evident the goal of the two checks: they are used exactly once each one and they are one-liners.

Reply:
> If you still think they need to be in a generic header file for everyone to use, I will make them stronger.

Now that we are on the same page I don't think that's necessary. We can keep them where they are, but as you indicated it's better to improve their naming.

Reply:
Done